src/freehand_selection.jl

```diff
@@ -1,61 +1,188 @@
         end
+        nothing
     end
 
-    append!(c.preserved, [sigstart, sigdrag, sigend])
+    push!(ctx.points, [])
+
+    append!(c.preserved, [sigstart, sigdraw, sigmove, sigend])
     Dict("enabled"=>enabled)
 end
 
+function near_vertex(pt::XY, p)
+    # returns index of nearby point
+    for i in 1:length(p)
+        if (p[i].x-5 <= pt.x <= p[i].x+5) && (p[i].y-5 <= pt.y <= p[i].y+5)
+            return i
+        end
+    end
+    return -1
+end
+
 function init_polygon_select(ctx::ImageContext)
-    enabled = Signal(true)
+    enabled = Signal(false)
     c = ctx.canvas
 
     dummybtn = MouseButton{UserUnit}()
-    push!(ctx.points, [])
+    building = Signal(false)
+    moving = Signal(false)
+    modifying = Signal(false)
+    num_pts = Signal(0)
+    modhandle = Signal(-1)
+    diff = Signal(XY(NaN,NaN))
+    mouse_motion = map(enabled,building) do e,b
+        #@show e,b
+        if e == true && b == true
+            true
+        else
+            false
+        end
+    end
 
     sigstart = map(filterwhen(enabled, dummybtn, c.mouse.buttonpress)) do btn
-        if ispolygon(value(ctx.points)) && !isinside(Point(btn.position),Point.(value(ctx.points)))
+        #println("Polygon enabled")
+        if (ispolygon(value(ctx.points)) &&
+            near_vertex(btn.position, value(ctx.points)) != -1)
+            # modification actions
+            #println("Modifying")
+            push!(modifying,true)
+            push!(building,false)
+            push!(moving,false)
+            push!(modhandle,near_vertex(btn.position,value(ctx.points)))
+        elseif (ispolygon(value(ctx.points)) &&
+                isinside(Point(btn.position),Point.(value(ctx.points))))
+            # moving actions
+            #println("Moving")
+            push!(moving,true)
+            push!(building,false)
+            push!(modifying,false)
+            push!(diff, XY(btn.position.x-value(ctx.points)[1].x,
+                           btn.position.y-value(ctx.points)[1].y))
+        elseif (ispolygon(value(ctx.points)) &&
+                !isinside(Point(btn.position),Point.(value(ctx.points))))
+            #println("Resetting")
+            push!(building,true)
+            # resets ctx.points
             push!(ctx.points,[])
-            Reactive.run_till_now()
-        end
-        if !ispolygon(value(ctx.points))
-            next = btn.position
-            if length(value(ctx.points)) > 3 # and click is near start
-                next = value(ctx.points)[1]
+            push!(num_pts, 0)
+        elseif isempty(value(ctx.points)) # adds first point
+            #println("First point")
+            push!(building,true)
+            push!(ctx.points, [btn.position])
+            push!(num_pts,1)
+        elseif !ispolygon(value(ctx.points)) # adds to polygon
+            if (length(value(ctx.points)) > 3 &&
+                value(ctx.points)[1].x - 5 <= btn.position.x <=
```

Comment:
would be good to avoid hard-coding `5` in more than one place. Can you call `near_vertex`? If not, perhaps define a `const` whose value is 5, and make use of the `const`. That way, if we ever decide this tolerance is too sloppy, it will change everywhere if we just change the `const`.